subprojects/base-services/src/main/java/org/gradle/internal/io/StreamByteBuffer.java

```diff
@@ -1,528 +1,529 @@
+            }
+            bytesLeft -= readBytes;
+        }
+    }
+
+    public void readFully(InputStream inputStream) throws IOException {
+        while (true) {
+            int len = allocateSpace();
+            int readBytes = currentWriteChunk.readFrom(inputStream, len);
+            if (readBytes == -1) {
+                break;
+            }
+        }
+    }
+
+    public byte[] readAsByteArray() {
+        byte[] buf = new byte[totalBytesUnread()];
+        input.readImpl(buf, 0, buf.length);
+        return buf;
+    }
+
+    public List<byte[]> readAsListOfByteArrays() {
+        List<byte[]> listOfByteArrays = new ArrayList<byte[]>(chunks.size() + 1);
+        byte[] buf;
+        while ((buf = input.readNextBuffer()) != null) {
+            if (buf.length > 0) {
+                listOfByteArrays.add(buf);
+            }
+        }
+        return listOfByteArrays;
+    }
+
+    public String readAsString(String encoding) {
+        Charset charset = Charset.forName(encoding);
+        return readAsString(charset);
+    }
+
+    public String readAsString() {
+        return readAsString(Charset.defaultCharset());
+    }
+
+    public String readAsString(Charset charset) {
+        try {
+            return doReadAsString(charset);
+        } catch (CharacterCodingException e) {
+            throw UncheckedException.throwAsUncheckedException(e);
+        }
+    }
+
+    private String doReadAsString(Charset charset) throws CharacterCodingException {
+        int unreadSize = totalBytesUnread();
+        if (unreadSize > 0) {
+            return readAsCharBuffer(charset).toString();
+        }
+        return "";
+    }
+
+    private CharBuffer readAsCharBuffer(Charset charset) throws CharacterCodingException {
+        CharsetDecoder decoder = charset.newDecoder().onMalformedInput(
+                CodingErrorAction.REPLACE).onUnmappableCharacter(
+                CodingErrorAction.REPLACE);
+        CharBuffer charbuffer = CharBuffer.allocate(totalBytesUnread());
+        ByteBuffer buf = null;
+        boolean wasUnderflow = false;
+        ByteBuffer nextBuf = null;
+        boolean needsFlush = false;
+        while (hasRemaining(nextBuf) || hasRemaining(buf) || prepareRead() != -1) {
+            if (hasRemaining(buf)) {
+                // handle decoding underflow, multi-byte unicode character at buffer chunk boundary
+                if (!wasUnderflow) {
+                    throw new IllegalStateException("Unexpected state. Buffer has remaining bytes without underflow in decoding.");
+                }
+                if (!hasRemaining(nextBuf) && prepareRead() != -1) {
+                    nextBuf = currentReadChunk.readToNioBuffer();
+                }
+                // copy one by one until the underflow has been resolved
+                buf = ByteBuffer.allocate(buf.remaining() + 1).put(buf);
+                buf.put(nextBuf.get());
+                ((Buffer)buf).flip();
```

Comment:
The line break of this file in repo is surprisingly `\r\n`. Since we're are using Unix system, it's converted to `\n`. Therefore you'd better change `Diff settings` in right-top corner of this page.